File: src/graph/load_error.zig
//! Typed load-error shape for untrusted surfaces (task-14 / C3).
//!
//! Mirrors `src/ts/graph/load_error.ts`. C2 load-error goldens consume
//! `code` (as `error_code`) and optional context (`offending`).

const std = @import("std");

/// Trust boundary phase (one of the four surfaces).
pub const Phase = enum {
    dsl,
    wire,
    manifest,
    graph_json,

    pub fn name(self: Phase) []const u8 {
        return switch (self) {
            .dsl => "dsl",
            .wire => "wire",
            .manifest => "manifest",
            .graph_json => "graph_json",
        };
    }
};

pub const Position = struct {
    byte: ?usize = null,
    line: ?u32 = null,
    col: ?u32 = null,
    /// Schema / JSON path (arena-owned when filled by callers).
    path: ?[]const u8 = null,
};

pub const Context = struct {
    node: ?[]const u8 = null,
    edge: ?[]const u8 = null,
    port: ?[]const u8 = null,
    field: ?[]const u8 = null,
};

/// Serializable failure description (not an error set — use with GraphError).
pub const LoadError = struct {
    phase: Phase,
    /// Stable code matching C2 goldens / Zig error names where applicable.
    code: []const u8,
    message: []const u8,
    position: Position = .{},
    context: Context = .{},

    pub fn offending(self: LoadError) ?[]const u8 {
        return self.context.node orelse self.context.edge orelse self.context.port orelse self.context.field;
    }
};

/// Thread-local last load error for surfaces that cannot return a struct through
/// Zig error unions (mirrors runner last_error pattern).
threadlocal var last_buf: [1024]u8 = undefined;
threadlocal var last_len: usize = 0;
threadlocal var last_phase: Phase = .graph_json;
threadlocal var last_code: [64]u8 = undefined;
threadlocal var last_code_len: usize = 0;

pub fn setLast(phase: Phase, code: []const u8, comptime fmt: []const u8, args: anytype) void {
    last_phase = phase;
    const clen = @min(code.len, last_code.len);
    @memcpy(last_code[0..clen], code[0..clen]);
    last_code_len = clen;
    const msg = std.fmt.bufPrint(&last_buf, fmt, args) catch blk: {
        const n = @min(fmt.len, last_buf.len);
        @memcpy(last_buf[0..n], fmt[0..n]);
        break :blk last_buf[0..n];
    };
    last_len = msg.len;
}

pub fn clearLast() void {
    last_len = 0;
    last_code_len = 0;
}

pub fn lastMessage() []const u8 {
    return last_buf[0..last_len];
}

pub fn lastCode() []const u8 {
    return last_code[0..last_code_len];
}

pub fn lastPhase() Phase {
    return last_phase;
}

File: src/ts/graph/load_error.ts
/**
 * Typed error model for untrusted-input surfaces (task-14 / C3).
 *
 * Shape shared with C2 load-error goldens (`error_code` ≡ `code`,
 * `offending` ≡ `context.node` / `context.edge`). Extend, don't fork.
 *
 *   { phase, code, position (byte/line-col or schema path), context, message }
 */

/** Trust boundary phase (one of the four surfaces). */
export type TrustSurfacePhase = "dsl" | "wire" | "manifest" | "graph_json";

/** Byte offset and/or line-col and/or JSON schema path. */
export type ErrorPosition = {
  /** 0-based byte (or UTF-16 code unit) offset into the source. */
  byte?: number;
  line?: number;
  col?: number;
  /** JSON / schema path, e.g. `nodes[3].expr` or `inputs[1].name`. */
  path?: string;
};

/** Offending structural location when known. */
export type ErrorContext = {
  node?: string;
  edge?: string;
  port?: string;
  field?: string;
};

/** Serializable load/parse failure (goldens + harnesses consume this). */
export type LoadErrorShape = {
  phase: TrustSurfacePhase;
  code: string;
  position?: ErrorPosition;
  context?: ErrorContext;
  message: string;
};

/**
 * Base typed error for all four untrusted surfaces.
 * `name` stays surface-specific on subclasses (DslError, WireDecodeError, …)
 * so existing `instanceof` / message checks keep working.
 */
export class MathZigLoadError extends Error {
  readonly phase: TrustSurfacePhase;
  readonly code: string;
  readonly position?: ErrorPosition;
  readonly context?: ErrorContext;

  constructor(
    phase: TrustSurfacePhase,
    code: string,
    message: string,
    opts?: { position?: ErrorPosition; context?: ErrorContext; cause?: unknown },
  ) {
    super(message);
    this.name = "MathZigLoadError";
    this.phase = phase;
    this.code = code;
    this.position = opts?.position;
    this.context = opts?.context;
    if (opts?.cause !== undefined) {
      (this as Error & { cause?: unknown }).cause = opts.cause;
    }
  }

  /** Golden / harness payload (C2-compatible fields derived). */
  toShape(): LoadErrorShape {
    return {
      phase: this.phase,
      code: this.code,
      position: this.position,
      context: this.context,
      message: this.message,
    };
  }

  /** C2 golden fields: error_code + offending. */
  toGoldenFields(): { error_code: string; offending?: string } {
    const offending =
      this.context?.node ?? this.context?.edge ?? this.context?.port ?? this.context?.field;
    return offending
      ? { error_code: this.code, offending }
      : { error_code: this.code };
  }
}

export class WireDecodeError extends MathZigLoadError {
  constructor(
    code: string,
    message: string,
    opts?: { position?: ErrorPosition; context?: ErrorContext; cause?: unknown },
  ) {
    super("wire", code, message, opts);
    this.name = "WireDecodeError";
  }
}

export class ManifestLoadError extends MathZigLoadError {
  constructor(
    code: string,
    message: string,
    opts?: { position?: ErrorPosition; context?: ErrorContext; cause?: unknown },
  ) {
    super("manifest", code, message, opts);
    this.name = "ManifestLoadError";
  }
}

export class GraphJsonError extends MathZigLoadError {
  constructor(
    code: string,
    message: string,
    opts?: { position?: ErrorPosition; context?: ErrorContext; cause?: unknown },
  ) {
    super("graph_json", code, message, opts);
    this.name = "GraphJsonError";
  }
}

/** Best-effort extract of LoadErrorShape from any thrown value. */
export function loadErrorShapeOf(err: unknown): LoadErrorShape | null {
  if (err instanceof MathZigLoadError) return err.toShape();
  if (err && typeof err === "object") {
    const e = err as Record<string, unknown>;
    if (typeof e.phase === "string" && typeof e.code === "string" && typeof e.message === "string") {
      return {
        phase: e.phase as TrustSurfacePhase,
        code: e.code,
        position: e.position as ErrorPosition | undefined,
        context: e.context as ErrorContext | undefined,
        message: e.message,
      };
    }
  }
  return null;
}
